File: round-1B/3-garbled/garbled.rb
#!/usr/bin/env ruby

$dictionary = Hash.new { |h, k| h[k] = [] }
$words = []
File.open("dictionary.txt").each_line do |word|
  word = word.chomp
  $dictionary[word[0]] << word
  $words << word
end
$cache = {}

def min_changes email, last_alter = 5, candidate = ''
  if email.empty?
    candidate.empty? ? 0 : Float::INFINITY
  elsif candidate.size > 0
    if email[0] == candidate[0] # word still match
      min_changes(email[1..-1], last_alter+1, candidate[1..-1])
    elsif last_alter >= 5 # a letter could have changed
      1 + min_changes(email[1..-1], 1, candidate[1..-1])
    else # don't match
      Float::INFINITY
    end
  else
    if $cache.has_key? [email, last_alter]
      return $cache.fetch([email, last_alter])
    end
    # puts "min_changes %20s (#{last_alter}), #{candidate}" % email
    min = Float::INFINITY
    $dictionary[email[0]].each do |candidate|
      sub_changes = min_changes email[1..-1], last_alter+1, candidate[1..-1]
      # puts "try with #{candidate} gives #{sub_changes}" if sub_changes < 1000
      min = sub_changes if sub_changes < min
      break if min == 0
    end
    if min > 0 and last_alter >= 5
      $words.each do |candidate|
        sub_changes = 1 + min_changes(email[1..-1], 1, candidate[1..-1])
        # puts "try with #{candidate} gives #{sub_changes}" if sub_changes < 1000
        min = sub_changes if sub_changes < min
        break if min == 1
      end
    end
    # puts "min_changes %20s (#{last_alter}), #{candidate} = #{min}" % email
    $cache[[email, last_alter]] = min
    min
  end
end

gets.to_i.times do |i|
  email = gets.chomp
  print "Case ##{i+1}: "
  $stdout.flush
  puts min_changes(email)
end

File: 2013/round-1B/3-garbled/garbled.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <cstring>

using namespace std;

typedef pair<string, int> cache_key;
map<char, vector<string> > dictionary;
vector<string> words;
map<cache_key, int> cache;
const int infinity = 1 << 30;

int min_changes(string email, int last_alter = 5, string candidate = "")
{
  if(email == "")
    return(candidate == "" ? 0 : infinity);
  else if(candidate != "")
  {
    if(email[0] == candidate[0]) // word still match
      return min_changes(email.substr(1), last_alter+1, candidate.substr(1));
    else if(last_alter >= 5) // a letter could have changed
      return 1 + min_changes(email.substr(1), 1, candidate.substr(1));
    else // don't match
      return infinity;
  }
  else
  {
    // cout << "min_changes " << email << " (" << last_alter << "), " << candidate << endl;
    if(cache[cache_key(email, last_alter)])
      return cache[cache_key(email, last_alter)];
    int min = infinity;
    vector<string> &indexed_words = dictionary[email[0]];
    for(int i = 0; i < indexed_words.size(); i++)
    {
      string &word = indexed_words[i];
      int sub_changes = min_changes(email.substr(1), last_alter+1, word.substr(1));
      if(sub_changes < min)
        min = sub_changes;
      if(min == 0)
        break;
    }
    if(min > 0 && last_alter >= 5)
    {
      for(int i = 0; i < words.size(); i++)
      {
        string &word = words[i];
        int sub_changes = 1 + min_changes(email.substr(1), 1, word.substr(1));
        if(sub_changes < min)
          min = sub_changes;
        if(min == 1)
          break;
      }
    }
    cache[cache_key(email, last_alter)] = min;
    return min;
  }
}


int main(int argc, char const *argv[])
{
  ifstream  dic_file;
  dic_file.open("dictionary.txt");
  for(int i = 0; i < 521196; i++)
  {
    string word;
    dic_file >> word;
    words.push_back(word);
    dictionary[word[0]].push_back(word);
  }

  if (argc >= 1) {
    ifstream  file;
    int       cases;

    file.open(argv[1]);
    file >> cases;
    for (int c = 0; c < cases; c++) {
      string email;
      file >> email;
      cout << "Case #" << c+1 << ": " << min_changes(email) << endl;
    }
    file.close();
    return 0;
  }
  return 1;
}